pyexchange/erisx.py

```diff
@@ -0,0 +1,399 @@
+
+        client_order_id = uuid.uuid4()
+        side = 1 if is_sell is False else 2
+        base_currency = pair.split('/')[0]
+
+        message.append_pair(simplefix.TAG_CLORDID, client_order_id)
+        message.append_pair(simplefix.TAG_HANDLINST, simplefix.HANDLINST_AUTO_PRIVATE)
+        message.append_pair(simplefix.TAG_CURRENCY, base_currency)
+        message.append_pair(simplefix.TAG_SIDE, side)
+        message.append_pair(simplefix.TAG_SYMBOL, pair)
+        message.append_pair(460, 2)
+        message.append_utc_timestamp(simplefix.TAG_TRANSACTTIME)
+        message.append_pair(simplefix.TAG_ORDERQTY, amount)
+        message.append_pair(simplefix.TAG_ORDTYPE, simplefix.ORDTYPE_LIMIT)  # always place limit orders
+        message.append_pair(simplefix.TAG_PRICE, price)
+        message.append_pair(simplefix.TAG_TIMEINFORCE, simplefix.TIMEINFORCE_GOOD_TILL_CANCEL)
+
+        #  Optional
+        message.append_pair(448, self.fix_trading_user)
+
+        self.fix_trading.write(message)
+        new_order = self.fix_trading.wait_for_response('8')
+
+        erisx_oid = new_order.get(simplefix.TAG_ORDERID).decode('utf-8')
+        client_oid = new_order.get(simplefix.TAG_CLORDID).decode('utf-8')
+        order_id = f"{erisx_oid}|{client_oid}"
+
+        return order_id
+
+    def cancel_order(self, order_id: str, pair: str, is_sell: bool) -> bool:
+        assert(isinstance(order_id, str))
+        assert(isinstance(pair, str))
+        assert(isinstance(is_sell, bool))
+
+        side = 1 if is_sell is False else 2
+        erisx_oid = order_id.split('|')[0]
+        client_oid = order_id.split('|')[1]
+        message = self.fix_trading.create_message(simplefix.MSGTYPE_ORDER_CANCEL_REQUEST)
+
+        message.append_pair(simplefix.TAG_CLORDID, uuid.uuid4())
+        message.append_pair(simplefix.TAG_ORDERID, erisx_oid)
+        message.append_pair(simplefix.TAG_ORIGCLORDID, client_oid)
+        message.append_pair(simplefix.TAG_SYMBOL, pair)
+        message.append_pair(simplefix.TAG_SIDE, side)
+        message.append_utc_timestamp(simplefix.TAG_TRANSACTTIME)
+
+        self.fix_trading.write(message)
+
+        response = self.fix_trading.wait_for_response('8')
+        return True if response.get(150).decode('utf-8') == '4' else False
+
+    # Trade information is only retrieved on a per session basis through FIX (Page 20 of Spec)
+    def get_trades(self, pair: str, page_number: int = 8) -> List[Trade]:
+        response = self._http_post("trades", {"account_id": self.account_id})
+        return list(map(lambda trade: ErisxTrade.from_message(trade), response["trades"]))
+
+    # used to unsubscribe from the marketdata feed used in get_all_tradesx
+    def unsubscribe_marketdata(self, pair: str, request_id: str) -> bool:
+        assert(isinstance(request_id, str))
+
+        message = self.fix_marketdata.create_message(simplefix.MSGTYPE_MARKET_DATA_REQUEST)
```

Comment:
Although I neglected to do so with `get_markets`, I'd recommend moving message construction to the `ErisxFix` class below.

Reply:
just for these marketdata methods, or for all methods that involve FIX message construction?

Reply:
After examining `place_order`, I'm having second thoughts about this.  Lots of arguments would need to be duplicated, complicating code maintenance.

Happy to leave it alone for now.  When we do the next integration, we'll more clearly see opportunities to move some `place_order` and `cancel_order` message construction logic to the base `FixEngine` class, and use `ErisxFix` to override the implementation.

Reply:
Yeah that makes sense to me